### PR TITLE
integration tests for old state

### LINKED ISSUE #108: Test old state support
Some RPC methods like `starknet_call` offer the user to perform a call at a historic block, by providing an old block specifier (by hash or number). Candidate methods include those that read the state and have `block_id` as their argument in the [RPC spec file](https://github.com/0xSpaceShard/starknet-devnet-rs/blob/main/crates/starknet-devnet-server/test_data/spec/0.7.1/starknet_api_openrpc.json): `starknet_call`, `starknet_estimateFee`, `starknet_estimateMessageFee`, `starknet_getStorageAt`, `starknet_getNonce`, `starknet_getClass`, `starknet_simulateTransactions`.

The old state support itself was introduced with #104, but not thoroughly tested. Currently we only test with `BlockId::Tag(BlockTag::Latest)`, but it would make sense to also test with `BlockId::Number(<HISTORIC_BLOCK_NUMBER>)` and `BlockId::Hash(<HISTORIC_BLOCK_HASH>)`

### PR DESCRIPTION
**Pull request by marioiordanov**
closes #108 
## Checklist:

- [ ] Checked out the [contribution guidelines](CONTRIBUTING.md)
- [ ] Applied formatting - `./scripts/format.sh`
- [ ] No linter errors - `./scripts/clippy_check.sh`
- [ ] No unused dependencies - `./scripts/check_unused_deps.sh`
- [ ] No spelling errors - `./scripts/check_spelling.sh`
- [ ] Performed code self-review
- [ ] Rebased to the latest commit of the target branch (or merged it into my branch)
    -   Once you make the PR reviewable, please avoid force-pushing
- [ ] Updated the docs if needed - `./website/README.md`
- [ ] Linked the [issues](https://github.com/0xSpaceShard/starknet-devnet-rs/issues) resolvable by this PR - [linking info](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)
- [ ] Updated the tests if needed; all passing - [execution info](https://github.com/0xSpaceShard/starknet-devnet-rs/blob/main/.github/CONTRIBUTING.md#test-execution)

### CODE REVIEW

File: crates/starknet-devnet/tests/test_minting.rs [hunk 1]
```diff
@@ -229,4 +230,33 @@ mod minting_tests {
...
+            let BlockHashAndNumber { block_hash, .. } =
+                devnet.json_rpc_client.block_hash_and_number().await.unwrap();
+
+            devnet.mint(address, mint_amount).await;
+
+            let balance_at_block =
+                devnet.get_balance_at_block(&address, BlockId::Hash(block_hash)).await.unwrap();
+            let latest_balance = devnet.get_balance_latest(&address, unit).await.unwrap();
+
+            assert!(balance_at_block + Felt::from(mint_amount) == latest_balance);
```
Comment (FabijanC) [inline crates/starknet-devnet/tests/test_minting.rs:259]: Perhaps better to use `assert_eq!` in this case

**Review by FabijanC — COMMENTED** [review]
If we're not going to add tests for other methods mentioned in #108, it can be approved.

File: crates/starknet-devnet/tests/test_minting.rs [hunk 2]
```diff
@@ -229,4 +230,33 @@ mod minting_tests {
...
+        let unit = FeeUnit::WEI;
+
+        for _ in 0..3 {
+            let BlockHashAndNumber { block_hash, .. } =
+                devnet.json_rpc_client.block_hash_and_number().await.unwrap();
+
+            devnet.mint(address, mint_amount).await;
+
+            let balance_at_block =
+                devnet.get_balance_at_block(&address, BlockId::Hash(block_hash)).await.unwrap();
```
Comment (FabijanC) [inline crates/starknet-devnet/tests/test_minting.rs:256]: Other than `starknet_getStorageAt`, perhaps also add old-block fee estimation, and old-block `starknet_call`, as described on Discord. Of course, if some of the existing tests don't already test this. In an ideal case, we would also test simulation at old block in a similar manner (basically all untested methods which are mentioned in #108).

Comment (marioiordanov) [reply crates/starknet-devnet/tests/test_minting.rs]: [43a7bea](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/615/commits/43a7beadef46c2a7edde7774fe0d487b20436350)

File: crates/starknet-devnet/tests/test_minting.rs (hunk 1 above)
Comment (marioiordanov) [reply crates/starknet-devnet/tests/test_minting.rs]: [43a7bea](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/615/commits/43a7beadef46c2a7edde7774fe0d487b20436350)

File: crates/starknet-devnet/tests/test_minting.rs (hunk 2 above)
Comment (FabijanC) [reply crates/starknet-devnet/tests/test_minting.rs]: Good additions to our tests. I browsed the existing tests once more and realized we already have a test which tests `getClass` at old blocks (in `test_get_class.rs`: `test_getting_class_at_various_blocks`). Should we leave both tests?

What we still don't have is a simple `call` test. Perhaps one of the existing tests you added could be modified to just make a call on ERC20 and get the balance.

Comment (marioiordanov) [reply crates/starknet-devnet/tests/test_minting.rs]: get_balance_at_block internally makes a starknet_call

Comment (FabijanC) [reply crates/starknet-devnet/tests/test_minting.rs]: Right, didn't see it.

File: crates/starknet-devnet/tests/test_old_state.rs [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+        let calldata = account.encode_calls(&calls);
+
+        let block_id = BlockId::Hash(block_hash);
+        let estimate_fee_error = devnet
+            .json_rpc_client
+            .estimate_fee(
+                [BroadcastedTransaction::Invoke(BroadcastedInvokeTransaction::V1(
+                    BroadcastedInvokeTransactionV1 {
+                        max_fee: Felt::ZERO,
+                        // precalculated signature
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-devnet/tests/test_old_state.rs]: [aa575bc](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/615/commits/aa575bc1530b29459e7094235c68e8ddd3378e54)

File: crates/starknet-devnet/tests/test_old_state.rs [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+            .unwrap_err();
+
+        let simulation_error = devnet
+            .json_rpc_client
+            .simulate_transaction(
+                block_id,
+                BroadcastedTransaction::Invoke(BroadcastedInvokeTransaction::V1(
+                    BroadcastedInvokeTransactionV1 {
+                        max_fee: Felt::from(1e18 as u128),
+                        // precalculated signature
```
[the remark this answers is not in the document]
Comment (marioiordanov) [reply crates/starknet-devnet/tests/test_old_state.rs]: [aa575bc](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/615/commits/aa575bc1530b29459e7094235c68e8ddd3378e54)

File: crates/starknet-devnet/tests/test_minting.rs (hunk 2 above)
Comment (marioiordanov) [reply crates/starknet-devnet/tests/test_minting.rs]: [2fbae41](https://github.com/0xSpaceShard/starknet-devnet-rs/pull/615/commits/2fbae415e3117b142d4a6c156e90679fda992813)

**Review by FabijanC — APPROVED** [review]

**Pull request merged**